### PR TITLE
DV2: Only run T+D if we have zero records or the previous sync left behind un-T+D-ed records

### PR DESCRIPTION
**Pull request by edgao**
Seeking early review: lmk if this interface makes sense. Bigquery+snowflake both implemented.

Non-draft to trigger CI.

Rough description:
* Add a `recordCounts` map to the OnCloseFunction interfaces, which tracks how many records were processed during this sync per stream.
* AsyncStreamConsumer - track this map
    * The `Executors.newFixedThreadPool(5)` thing is copied from [this](https://github.com/airbytehq/airbyte/blob/f1b91b3b4bce79576a08a519048fb7bfb45400ef/airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/FlushWorkers.java#L75) FlushWorkers constructor, to avoid duplicate constructor code
* BufferedStreamConsumer - don't bother fixing this class, we want to kill it anyway
* Pass `recordCounts` into the appropriate TyperDeduper method
* Modify the DestinationHandler interface to find more detailed information at the start of a sync (we were already computing this)

failing ci is unrelated, investigation [ongoing](https://airbytehq-team.slack.com/archives/C03VDJ4FMJB/p1701977010235689)
```
BigQueryStandardInsertsTypingDedupingTest > testRawTableJsonToStringMigration() FAILED
    io.airbyte.workers.exception.TestHarnessException: Could not find image: airbyte/destination-bigquery:1.9.0
```

### CODE REVIEW

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java [hunk 1]
```diff
@@ -204,7 +204,7 @@ public void close() throws Exception {
     flushWorkers.close();
 
     bufferManager.close();
-    onClose.accept(hasFailed);
+    onClose.accept(hasFailed, recordCounts);
```
[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java:207]: I like that! will make the change :+1:

[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java]: how does https://github.com/airbytehq/airbyte/pull/33232/commits/3a06afc32460bfdbbba3e3edb500225fc1576962 look?

[the remark this answers is not in the document]
Comment (gisripa) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java]: lgtm. _resolving_

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java [hunk 2]
```diff
@@ -204,7 +210,14 @@ public void close() throws Exception {
     flushWorkers.close();
 
     bufferManager.close();
-    onClose.accept(hasFailed);
+
+    final Map<StreamDescriptor, StreamSyncSummary> streamSyncSummaries = streamNames.stream().collect(toMap(
+        streamDescriptor -> streamDescriptor,
+        streamDescriptor -> new StreamSyncSummary(
+            Optional.of(recordCounts.get(streamDescriptor)).map(AtomicLong::get)
```
Comment (gisripa) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java:217]: Won't this work ? Optional.of(recordCounts.get(streamDescriptor).get()) I see you already defaulted it to AtomicLong(0) in constructor. so not present wouldn't be the case ?

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java:217]: lol I was waffling between "preinitialize in the constructor" and "defensive handling when consuming". Yes, I should remove one of those - I'm inclined to keep the defensive handling here (+ switch it to a getOrDefault) and remove the preinit in the constructor, wdyt?

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java]: pushed https://github.com/airbytehq/airbyte/pull/33232/commits/e24c5123069acba45a94cdfa3712aef8d0565ad8.

(also... my previous code wasn't even safe. It should have been Optional.ofNullable, at least >.>)

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java]: ok this should _actually_ work now https://github.com/airbytehq/airbyte/pull/33232/commits/740ee1af634b1f9032c31ff741abf0eecbc1b927

(I always mix up the map get+default methods >.>)

**Comment by edgao** [comment]
can't repro the test failure locally 🤔 so, uh, idk

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java (hunk 2 above)
Comment (gisripa) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java]: 👍

**Review by gisripa — APPROVED** [review]
lgtm _pending logistics_

**Comment by edgao** [comment]
agh I need to publish the cdk here >.> will literally never remember to do this

**Comment by edgao** [comment]
... which also means I can't merge this until https://github.com/airbytehq/airbyte/pull/33369 is merged :cry:

**Comment by gisripa** [comment]
https://github.com/airbytehq/airbyte/pull/33369 is merged

**Pull request merged**